```{r echo=FALSE, message=FALSE, warning=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("Gviz")
BiocManager::install("GenomicRanges")
install.packages("sqldf")
```

```{r echo=FALSE, message=FALSE, warning=FALSE}
library(Gviz)
library(GenomicRanges)
library(BSgenome)
library(Biostrings)
library(rtracklayer)
library(sqldf)
```



# Working example with sea lion genome

## Genome Track
```{r}
options(ucscChromosomeNames = FALSE)
getOption("Gviz.scheme")
data(collapseTrack)

plot_start <- 1
plot_end <- 3986568

chr <- "CM019820.2"

genometrack <- GenomeAxisTrack()

gen <- readDNAStringSet("/home/yakupova/1PROJECTS/Y/assembly/Y.fa")
```

## Sequence Track
```{r}
# sequenceTrack <- SequenceTrack(gen, genome = "ZalCal", chromosome = chr, name = "Sequence")
```

## SNP Track (MassArray)
```{r}
snp_file <- "/home/yakupova/1PROJECTS/Y/haplotype_analysis/MassArraySNPpositions.txt"
snp_data <- read.csv(snp_file, sep = "\t")
snpTrackMassArray <- AnnotationTrack(
  start = snp_data$Position,
  end = snp_data$Position,
  chromosome = chr,
  genome = gen,
  id = snp_data$Assay,
  feature = snp_data$Assay,
  name = "MassArray SNPs",
  stacking = "dense"
)
```

## SNP Track (Bioinf)
```{r}
snp_file <- "/home/yakupova/1PROJECTS/Y/Y_annotation_plot/All_males_CSL_Y_snp.bed"
snp_data <- read.csv(snp_file, sep = "\t")
snpTrackBioinf <- AnnotationTrack(
  start = snp_data$Position,
  end = snp_data$Position,
  chromosome = chr,
  genome = gen,
  name = "Bioinf SNPs",
  stacking = "dense"
)
```
## Annotation Track
```{r}
Yregion_col <- c(
  "pseudo" = "mediumseagreen",
  "single" = "steelblue3",
  "multi" = "#fdf065",
  "Xtrans" = "palevioletred2"
)

annotation_file <- "/home/yakupova/1PROJECTS/Y/Y_annotation_plot/Ychromsome_annotation_Rin.csv"
anno <- read.csv.sql(annotation_file, sep = '\t', sql="select * from file where gene_biotype != 'pseudogene'")
geneTrack <- AnnotationTrack(
            genome = "ZalCal", name = "Genes", chromosome = chr,
            start = anno$Start, end = anno$Stop, strand = anno$Strand,
            feature = anno$Name,
            fill = Yregion_col[anno$Yregion],
            col = "white",
            stacking="hide", shape="fixedArrow", arrowHeadWidth=100)
```

## Homologous Track
```{r}
homolog_file <- "/home/yakupova/1PROJECTS/Y/blast/homologous_regions.rX_wo_PAR_cumsum.qY.filtered1000.90pident.sorted.merged.bed"
homolog <- read.csv(homolog_file, sep = '\t')
homologTrack <- AnnotationTrack(
                genome = "ZalCal", name = "Homologs", chromosome = chr,
                start = homolog$Start, end = homolog$End,
                stacking="dense",
                col="#2180ce", fill="#2180ce")
```

## Coverage Track
```{r}
bam <- "/home/yakupova/1PROJECTS/Y/bam/G465_merged_duprm_subsample0_25_realigned.Y.bam"
coverageTrack <- DataTrack(range = bam, genome = "ZalCal", type = "l",
                           name = "Coverage", window = -1,
                           chromosome = chr, ylim = c(0, 20))
```

## Plot Tracks
```{r}
plotTracks(list(genometrack, geneTrack, snpTrackMassArray, snpTrackBioinf, homologTrack, coverageTrack),
           from = plot_start, to = plot_end,
           chromosome = chr,
           legend = TRUE, groupAnnotation="feature",
           rotation.group = 90, just.group = "above",
           background.panel = "white", background.title = "white",
           col.axis = "black", col.title = "black")
```
