```{r echo=FALSE, message=FALSE, warning=FALSE}
# if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")

# BiocManager::install("Gviz")
# BiocManager::install("GenomicRanges")
# install.packages("sqldf")
```

```{r echo=FALSE, message=FALSE, warning=FALSE}
library(Gviz)
library(GenomicRanges)
library(BSgenome)
library(Biostrings)
library(rtracklayer)
library(sqldf)
library(dplyr)
```


# Working example with sea lion genome

## Genome Track
```{r}
options(ucscChromosomeNames = FALSE)
getOption("Gviz.scheme")
data(collapseTrack)

plot_start <- 1
plot_end <- 3986568

gen <- readDNAStringSet("/home/yakupova/1PROJECTS/Y/assembly/Y.fa")
chr <- "CM019820.2"
```

## Genome Axis Track
```{r}
genometrack <- GenomeAxisTrack()
```

## Sequence Track
```{r}
# sequenceTrack <- SequenceTrack(gen, genome = "ZalCal", chromosome = chr, name = "Sequence")
```

## SNP Track (MassArray)
```{r}
snp_file <- "/home/yakupova/1PROJECTS/Y/haplotype_analysis/MassArraySNPpositions.txt"
snp_data <- read.csv(snp_file, sep = "\t")

snpTrackMassArray <- AnnotationTrack(genome = gen, chromosome = chr, name = "MassArray SNPs",
                                     start = snp_data$Position, end = snp_data$Position,
                                     id = snp_data$Assay, feature = snp_data$Assay,
                                     stacking = "dense", size = 0.2,
                                     col = "lightgrey", fill = "lightgrey"
)
```

## SNP Track (Bioinf)
```{r}
snp_file <- "/home/yakupova/1PROJECTS/Y/Y_annotation_plot/All_males_CSL_Y_snp.bed"
snp_data <- read.csv(snp_file, sep = "\t")

snpTrackBioinf <- AnnotationTrack(genome = gen, chromosome = chr, name = "Bioinf SNPs",
                                  start = snp_data$Position, end = snp_data$Position,
                                  stacking = "dense", size = 0.2,
                                  col = "lightgrey", fill = "lightgrey"
)
```
## Genes Track
```{r}
Yregion_col <- c(
  "pseudo" = "mediumseagreen",
  "single" = "steelblue3",
  "multi" = "#fdf065",
  "Xtrans" = "palevioletred2"
)

annotation_file <- "/home/yakupova/1PROJECTS/Y/Y_annotation_plot/Ychromsome_annotation_Rin.csv"
anno <- read.csv.sql(annotation_file, sep = "\t", sql = "select * from file where gene_biotype != 'pseudogene'")

geneTrack <- AnnotationTrack(genome = "ZalCal", chromosome = chr, name = "Genes",
                             start = anno$Start, end = anno$Stop, strand = anno$Strand,
                             feature = anno$Name, groups = anno$Yregion,
                             fill = Yregion_col[anno$Yregion], col = "transparent",
                             stacking = "hide", shape = "fixedArrow", arrowHeadWidth = 100,
                             size = 0.5
)
```

## Repeats Track
```{r}
repeat_class_col <- c(
  "DNA" = "steelblue3",
  "LINE" = "mediumseagreen",
  "LTR" = "#fdf065",
  "SINE" = "palevioletred2",
  "RNA" = "orange",
  "Satellite" = "#c02727",
  "Simple_repeat" = "lightgrey",
  "Low_complexity" = "lightgrey"
)
```

```{r}
# Function creates track for each type/class of repeats or any other feature with name in DataFrame df
make_track <- function(name, df) {
  df <- df %>% filter(class %in% c(name))
  track <- AnnotationTrack(genome = genome, chromosome = chr, range = df,
                           name = df$class[1], groups = df$class,
                           fill = repeat_class_col[df$class],
                           col = "transparent",
                           stacking = "dense", size = 0.1
                          )
  return(track)
}
```

```{r}
repeats_file <- "/home/yakupova/1PROJECTS/Y/Y_annotation_plot/GCA_009762305.2_mZalCal1.pri.v2_rm.CM019820.2.csv"
repeats <- read.csv(repeats_file, sep = '\t')
# repeats <- repeats %>% filter(!repeat_class %in% c("LINE/CR1", "LINE/L1", "LINE/L2", "LINE/Penelope", "LINE/RTE-BovB"))
# repeats <- repeats %>% filter(grepl("SINE", repeat_class))
```

```{r}
# Group subclasses of repeats into classes
repeats <- repeats %>%
  mutate(repeat_type = ifelse(grepl("LINE", repeat_class), "LINE",
                       ifelse(grepl("SINE", repeat_class), "SINE",
                       ifelse(grepl("LTR", repeat_class), "LTR",
                       ifelse(grepl("DNA", repeat_class), "DNA",
                       ifelse(grepl("RNA", repeat_class), "RNA", repeat_class))))))
```

```{r}
# Transform into GRanges representaion to merge overlapping ranges
gr_repeats <- GRanges(seqnames = repeats$query_sequence,
              ranges = IRanges(start = repeats$pos_in_query_begin, end = repeats$pos_in_query_end),
              class = repeats$repeat_type,
              # strand = ifelse(repeats$direction == "C", "*", repeats$direction)
              )
gr_repeats_reduced <- unlist(reduce(split(gr_repeats, gr_repeats$class)))
```

```{r}
# Go back to DataFrame represenation
repeats_reduced <- data.frame(
  seqnames = as.character(seqnames(gr_repeats_reduced)),
  start = start(gr_repeats_reduced),
  end = end(gr_repeats_reduced),
  class = names(gr_repeats_reduced)
)
```

```{r}
# Create track that will gather all subtracks of repeats under one track
repeatsTrack <- HighlightTrack(trackList = list(make_track("DNA", repeats_reduced),
                                                make_track("RNA", repeats_reduced),
                                                make_track("LINE", repeats_reduced),
                                                make_track("SINE", repeats_reduced),
                                                make_track("LTR", repeats_reduced)),
                               name = "Repeats")
```

## Homologous Track
```{r}
homolog_file <- "/home/yakupova/1PROJECTS/Y/blast/homologous_regions.rX_wo_PAR_cumsum.qY.filtered1000.90pident.sorted.merged.bed"
homolog <- read.csv(homolog_file, sep = "\t")

homologTrack <- AnnotationTrack(genome = "ZalCal", chromosome = chr, name = "Homologs",
                                start = homolog$Start, end = homolog$End,
                                stacking = "dense", size = 0.2,
                                col = "#2180ce", fill = "#2180ce"
)
```

## Coverage Track
```{r}
bam <- "/home/yakupova/1PROJECTS/Y/bam/G465_merged_duprm_subsample0_25_realigned.Y.bam"

coverageTrack <- DataTrack(range = bam, genome = "ZalCal", chromosome = chr, name = "Coverage\t(M)",
                            col = "#464343",
                            type = "l", window = -1, ylim = c(0, 20), size = 0.5
)
```

## Plot Tracks
```{r}
plotTracks(list(
              genometrack,
              geneTrack,
              snpTrackMassArray,
              snpTrackBioinf,
              repeatsTrack,
              homologTrack,
              coverageTrack),
           from = plot_start, to = plot_end,
           chromosome = chr,
           legend = TRUE, groupAnnotation = "feature",
           rotation.group = 90, just.group = "above",
           background.panel = "white", background.title = "white",
           col.axis = "black", col.title = "black")
```
```{r}
# Function to close all open connections
close_all_connections <- function() {
  open_conns <- showConnections(all = TRUE)
  for (conn in seq_len(nrow(open_conns))) {
    try(close(getConnection(as.integer(open_conns[conn, "description"]))), silent = TRUE)
  }
}

# Call the function to close all connections
close_all_connections()
```
